### ISSUE TITLE
Cross-references to hidden elements

### ISSUE THREAD

**Issue by peterwebster** [peterwebster/henson#116]
Hi @nomoregrapes : we have yet another conundrum to solve at two minutes to midnight: annotations that cross-reference to another person (usually) who is set to be hidden/not indexed.

It seems to me we have four possible options:

1. More dev work such that, when the application loads an annotation with a cross-ref in it, it first checks whether there is any data to load, and if not, does not print the 'See also:' 
This strikes me as involved, and quite a lot of work at this stage.

**Comment by peterwebster** [comment]
2. The TSV data is fixed, such that cross-references that don't yet point to anything are formatted as something like:  person.82.hide, and so the stylesheet knows without additional requests what to do.

**Comment by peterwebster** [comment]
3. Option 3 is to comment out *all* the functionality involving the cross-refs until the edition is complete (and the data is thus all there). but I'm not sure how we would go about revealilng that functionality without your (or someone's) dev support.

Or 4. change the error messaging, so that if user follows a link to a non-existent person/place, it says something like 'Bear with us, it's a work in progress, we'll add this in in due course.'

**Comment by peterwebster** [comment]
So @nomoregrapes concentrate right now on the server as agreed in #114 but we'll need to think about this at some point; thoughts welcome on the work involved in each option.

**Comment by nomoregrapes** [comment]
I don't think this will be an issue.

1. When the TSV ingests person 81 it will see a cross ref to person.82 and try to find that so it can make a 2-way data link. If it doesn't then it forgets it has a cross ref and so won't list any.
2. Later it is ingesting person 82 (usually the next row in the TSV) which would hopefully have a cross ref to person.81. It finds that and makes a 2-way data link.

If person 82 is in the TSV with no-index, then it will skip step 2 completely.
If person 82 becomes important, then a TSV with them needs to be ingested.

I've assumed cross-references should always be 2-way(mutual).
Regardless of no-index, there is a potential issue that if the later person does not reference an earlier-imported person then the link is not made in either direction. It's possible to fix that by re-running the linking part of the ingest, which I'll explain in my technical documentation. Some of my technical to-dos might have the side-affect of making it easier to rerun these processes.

**Comment by peterwebster** [comment]
Thanks @nomoregrapes : at the moment I'm relieved to think that we won't have links that go nowhere; more comfortable with links that don't exist in either direction.
@DurHHHI and I will need to find a few of these and test the behaviour once we've got the finalised data ingested.

**Comment by peterwebster** [comment]
@DurHHHI but the last part of Gregory's comment does give me pause:  are we right to assume that cross-references are always two-way? Or, could they be made to be without doing violence to the meaning of anything?

**Comment by DurHHHI** [comment]
Hi @peterwebster and @nomoregrapes 

In current Hilary and Katie cross-ref thinking, they are not always! For example, Dean's Yard ID (places) lists cross ref links to all the person IDs who lived there. We've not followed through with a places cross ref for each individual persons ID. Family too: people like Ella, H's wife, are only cross-ref'd to her immediate family but other people (extended family) are linked to her to help better understand the connection.

Does it always have to be a 2-way mutual agreement? Hoping not!

**Comment by peterwebster** [comment]
@nomoregrapes in light of Hilary's comment, how easy would it be to undo the mandatory two-way nature of the relationship? If you can give me a sense of the work involved, we can take a view on if/how to prioritise it alongside everything else in the run-in.
We didn't spell out whether it should or should not be so, and you made a reasonable working assumption in that absence.
I think it needs to allow:
(i) person.1 to person.2 where the link is reciprocal
(ii) person.1 to person.2 where link is not reciprocal
(iii) person.1 to person.2 where link is real but person.2 does not yet have an annotation [the original point of the ticket]

You may tell me that it's too late for this.

**Comment by peterwebster** [comment]
@nomoregrapes it may be related to whether or not all the TSV is reingested each time, as raised at #118

**Comment by nomoregrapes** [comment]
It's possible to stop assuming links should be reciprocal.

The trouble is that when ingesting person.1, neither (i) or (ii) can happen because person.2 hasn't been ingested yet.
#104 will need to be done first, this will allow the system to ingest all the people it has, and then do a delayed task of looking at the cross references.

**Comment by peterwebster** [comment]
OK @nomoregrapes let's come back to this once we've got a functioning application in its final home.

**Comment by nomoregrapes** [comment]
I just spotted, I've done most of the work already to make cross references one-directional.
I've disabled the code that jumps to assuming that a reference should be reciprocated.

Some stuff will need to be done to ensure all intended references are linked up, and potentially the existing links cleared out. That's all work that will be completed int he course of other tasks anyway.

**Comment by peterwebster** [comment]
@KPalmerHeathman @DurHHHI 
this is ready for testing, on the following points:

(i) test that cross-references to annotations set to index=no are not visible


ON TESTING: we think this is OK, but see new issue below.

**Comment by KPalmerHeathman** [comment]
http://10.245.28.108:3000/people/person112 
Isabella is cross referenced to five people, but far more are appearing. All of these people have cross refs to her, but should not be in her own refs.

**Comment by nomoregrapes** [comment]
I've confirmed this was due to having old cross refs (back when I imported them bi-directional), which I have now deleted from the database.

I'm in the process of reimporting the cross reference data for different items, I discovered a few code bugs along the way.

Not all people are getting ingested, and I need to figure out why this is.

**Comment by nomoregrapes** [comment]
This is now fixed.
[Isabella](http://10.245.28.108:3000/people/person112) only has 5 cross-refs. All cross-refs around the site should be correct and not have extras.

**Comment by peterwebster** [comment]
Could @DurHHHI @KPalmerHeathman check this one out, and close the ticket if you're happy?

**Comment by KPalmerHeathman** [comment]
The cross refernces seem to be behaving as they should, so I'm closing this now.
